webmin/os-eol-lib.pl

```diff
@@ -0,0 +1,176 @@
+};
+&http_download('endoflife.date', 443, "/api/$os.json", \$fetch, \$error, undef, 1);
+if ($error) {
+        &error_stderr("Could not fetch current OS EOL data: " . $error);
+        $eol_write_cache->('[]');
+        return undef;
+        }
+my $fetch_json;
+eval { $fetch_json = &convert_from_json($fetch); };
+if ($@) {
+        &error_stderr("Could not parse fetched OS EOL data: $@");
+        $eol_write_cache->('[]');
+        return undef;
+        }
+$eol_write_cache->($fetch);
+return $fetch_json;
+}
+
+# eol_get_os_data()
+# Returns the current OS EOL data from the cache or fetches it
+# if it is not available
+sub eol_get_os_data
+{
+my $eol_cache_file = "$module_var_directory/eolcache";
+my $eol_cache = &read_file_contents($eol_cache_file);
+if ($eol_cache) {
+        # Check if the cache is still valid (5 days)
+        if (time() - (stat($eol_cache_file))[9] < 60*60*24*5) {
+                eval { $eol_cache = &convert_from_json($eol_cache); };
+                if ($@) {
+                        unlink($eol_cache_file);
+                        &error_stderr("Could not parse current OS EOL data: $@");
+                        return undef;
+                        }
+                return (ref($eol_cache) eq 'ARRAY' && @$eol_cache) ?
+                        $eol_cache : undef;
+                }
+        }
+# No cache found, fetch the data
+my $eol_fetched = &eol_fetch_os_data();
+return $eol_fetched;
+}
+
+# eol_get_os_eol_data()
+# Returns the EOL hash for the current OS or undef if the OS is not supported
+sub eol_get_os_eol_data
+{
+my $os_version = lc($gconfig{'real_os_version'});
+my $os_type = $gconfig{'real_os_type'};
+my $os_version_formatter = sub {
+        my $v = shift;
+        # Extract the major and minor versions
+        $v =~ m/^(?:stable\/)?(?<major>\d+)(?:\.(?<minor>\d+))?/;
+        $v = $+{minor} ? "$+{major}.$+{minor}" : $+{major};
+        return undef if (!$+{major});
+        # Minor versions in cycle are allowed only for Ubuntu
+        return $+{major} if (lc($os_type) !~ /ubuntu/);
+        return $v;
+};
+$os_version = $os_version_formatter->($os_version);
+# Get the EOL data for the current OS
+my $eol_data = &eol_get_os_data();
+return undef if (!$eol_data);
+# Find the EOL data for the current OS version
+($eol_data) = grep { $_->{'cycle'} eq $os_version } @$eol_data;
+return undef if (!$eol_data);
+
+# Add OS real name
+$eol_data->{'_os'} = $os_type;
+
+# Convert EOL date to a timestamp and a human-readable date based on locale
+my ($year, $month, $day) = split('-', $eol_data->{'eol'});
+$month -= 1;
+my $eol_timestamp = timelocal(0, 0, 0, $day, $month, $year);
+my $eol_date = &make_date($eol_timestamp, { '_' => 1 });
+$eol_data->{'_eol'} =
+        { daymonth => $eol_date->{'complete_short'},
+          month => $eol_date->{'monthfull'},
+          year => $eol_date->{'year'},
+          short => $eol_date->{'short'},
+          timestamp => $eol_timestamp };
+$eol_data->{'_eol_in'} =
```

Comment:
Why not just return a timestamp and have the caller of this API format it as they want?

Reply:
It's better call it once and make object just ready.

Reply:
An internal API like this shouldn't contain a lot of presentation logic - it should be up to the caller to format the dates for display.

Reply:
Agreed! I have factored this logic out. Nevertheless, there is still a lot to do to prevent multiple callers from having to reimplement this logic - there has to be a dedicated subroutine for this.